TwoLevelSelector component

tumentumurchudur opened the pull request:
TwoLevelSelector component: 
- Removed MX specific functionality

  
  Summary of props
- data (object) - accepts an object in the following format:

  {
   'option1': ['sub11', 'sub12'],
   'option2': ['sub21', 'sub22', 'sub23'],
   'option3': ['sub31'],
   'option4': ['sub41', 'sub42', 'sub43', 'sub44', 'sub45', 'sub46', 'sub47', 'sub48'],
   'option5': ['sub51', 'sub52', 'sub53', 'sub54', 'sub55']
  }
- onChange(func) - callback function that passes the selected text
- title (string) - text for the input field before an option is selected. It is 'Choose an item' by default.
- selectedItem (string) - is the default selected text  

<img width="626" alt="screen shot 2015-11-04 at 10 36 54 am" src="https://cloud.githubusercontent.com/assets/9920303/10946619/6eace71c-82e0-11e5-9196-f3d67a5f51c4.png">

bsbeeks commented on src/components/TwoLevelSelector.js:
```diff
@@ -0,0 +1,377 @@
+const Icon = require('./Icon');
+const StyleConstants = require('../constants/Style');
+
+const TwoLevelSelector = React.createClass({
```
I'm not a huge fan of this name. Any other suggestions? Here are some options: `FlyoutMenu`, `TieredMenu`

cerinman replied:
I like FlyoutMenu

tumentumurchudur replied:
Sounds good. I will change the name to FlyoutMenu

bsbeeks commented on src/components/TwoLevelSelector.js:
```diff
@@ -0,0 +1,377 @@
+const TwoLevelSelector = React.createClass({
+
+  propTypes: {
+    data: React.PropTypes.object,
```
I think the expected data structure can be solidified a bit more. Here is what I propose:
```javascript
[
  {
    label: 'Has A Second Level Menu', //optional: if not set, just display the value
    value: 'menu-1',
    items: [
      {
        label: 'Second Level Menu Item', //optional: if not set, just display the value
        value: 'second-level-1',
      }
    ],
  },
  {
    label: 'No Second Level Menu', //optional: if not set, just display the value
    value: 'menu-2'
  }
]
```

bsbeeks commented on src/components/TwoLevelSelector.js:
```diff
@@ -0,0 +1,377 @@
+
+  getDefaultProps () {
+    return {
+      data: null,
```
It seems like the data prop should be required for this component to work. If that's the case, then we shouldn't set a default prop.

tumentumurchudur replied:
Yes, that is a good idea to require this prop.

cerinman commented on src/components/TwoLevelSelector.js:
```diff
@@ -0,0 +1,377 @@
+  propTypes: {
+    data: React.PropTypes.object,
+    onChange: React.PropTypes.func,
+    selectedItem: React.PropTypes.string,
```
So, as  a user of this component, I might want more than just a string back.  Sometimes I care more about the associated id or guid of a selected item rather than it's display name.  Can we change this and the code around it to pass back more than a string.  This might require changing how the data prop is formatted but that might make this component a little more usable. 

I imagine the data prop looking like this

```javascript
[
  {
    name: 'groceries'
    id: 1232341234
    subs: [
      {
        //same layout as parent
      }
    ]
  }
]
```

tumentumurchudur replied:
@cerinman @bsbeeks 
Yes, agreed. How about we change the data structure as follows:
```javascript
[
  {
    id: 1,
    value: 'groceries',
    items: [
      {
        id: 11,
        value: 'item1'
      },
      {
        id: 12,
        value: 'item2'
      }
    ]
  },
  {
    id: 2,
    value: 'books',
    items: [
      {
        id: 21,
        value: 'book1'
      }
    ]
  }
]
```

And then return an object onSelect event

cerinman replied:
I still think @bsbeeks format is better.  I like the idea of the label key containing the displayed text and the value key containing an id or guid.

tumentumurchudur replied:
Okay, I will change it. The structure is the same but the naming of properties is little different.
```javascript
[
  {
    label: 'groceries',
    value: '1xxxx',
    items: [
      {
        label: 'milk',
        value: '12xxx'
      },
      {
        label: 'eggs',
        value: '23xxx'
      }
    ]
  },
  {
    label: 'books',
    value: '31xxx',
    items: [
      {
        label: 'bookXY',
        value: '31xxx'
      }
    ]
  }
]
```

cerinman commented on src/components/TwoLevelSelector.js:
```diff
@@ -0,0 +1,377 @@
+  },
+
+  _handleClick (selectedItem) {
+    this.props.onChange(selectedItem);
```
And make sure the selectedItem you are returning is the entire object and not just the label as a string.

```javascript
{ label: 'groceries', value: 1 }
```

tumentumurchudur replied:
@cerinman 
Yup, I will make sure an object is returned from the array.

tumentumurchudur commented:
- Changed the data structure for the data prop
- Cleaned up the styles
- Changed the onSelect callback to return an object from the passed data array

<img width="514" alt="screen shot 2015-11-06 at 2 54 07 pm" src="https://cloud.githubusercontent.com/assets/9920303/11009747/b92a3bf2-8496-11e5-9e61-3be7ae0b1d2a.png">
<img width="281" alt="screen shot 2015-11-06 at 2 55 11 pm" src="https://cloud.githubusercontent.com/assets/9920303/11009748/bcc0af80-8496-11e5-8056-e65aebbe8596.png">

bsbeeks commented:
It doesn't look like we'll be using this anytime soon. We can reopen if needed.
